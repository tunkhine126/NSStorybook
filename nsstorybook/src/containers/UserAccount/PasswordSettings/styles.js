import { makeStyles } from '@material-ui/core/styles';
import styled from 'styled-components';

export const ConfirmPwdStyle = styled.div`
  .MuiFormControl-marginNormal {
    margin-top: 0;
  }
`;

export const pwdSettingsStyles = makeStyles(theme => ({
  confirmContainer: {
    paddingTop: '25px',
  },
  newPassword: {
    minHeight: '202px',
    borderBottom: '1px solid #BBBBBB',
    [theme.breakpoints.down('md')]: {
      minHeight: '225px',
    },
  },
}));
