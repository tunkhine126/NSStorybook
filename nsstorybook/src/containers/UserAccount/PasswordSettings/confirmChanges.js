/**
 *
 * NewPassword
 *
 */

import React from 'react';
import { FormattedMessage } from 'react-intl';

import Grid from '@material-ui/core/Grid';

import InputValidation from 'components/shared/InputValidation';
import { required } from 'utils/validations';
import messages from '../messages';
import { commonStyles } from '../styles';
import { ConfirmPwdStyle } from './styles';

export function ConfirmChanges() {
  const classes = commonStyles();

  return (
    <ConfirmPwdStyle>
      <Grid item className={classes.spacing}>
        <InputValidation
          field="currentPassword"
          type="password"
          label={<FormattedMessage {...messages.currentPassword} />}
          validate={required}
          validateOnChange
          validateOnBlur
        />
      </Grid>
    </ConfirmPwdStyle>
  );
}

export default ConfirmChanges;
