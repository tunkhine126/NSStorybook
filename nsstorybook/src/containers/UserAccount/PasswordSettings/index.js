/**
 *
 * PasswordSettings
 *
 */

import React, { useRef, useState } from 'react';
import { injectIntl, intlShape } from 'react-intl';
import { Form } from 'informed';
import { useMutation } from '@apollo/react-hooks';

import Grid from '@material-ui/core/Grid';

import Content from 'components/shared/Content';
import ActionPanel, { loadingAction } from 'components/shared/ActionPanel';
import ViewHeader from 'components/shared/ViewHeader';
import Notification from 'components/shared/Notification';
import Loading from 'components/shared/Loading';

import { UPDATE_PASSWORD } from 'graphql/mutations/user-mutation';
import { generalErrorHandler } from 'utils/error-handler';
import globalMessages from 'messages';
import NewPassword from './newPassword';
import ConfirmChanges from './confirmChanges';
import messages from '../messages';

import { commonStyles, InputWrapper } from '../styles';
import { pwdSettingsStyles } from './styles';

export function PasswordSettings({ intl, ...rest }) {
  const formApiRef = useRef();
  const classes = commonStyles();
  const pwdSettingsClasses = pwdSettingsStyles();

  const [loadingState, setloadingState] = useState(false);
  const [messageState, setmessageState] = useState({
    message: '',
    open: false,
    type: 'success',
  });

  const [updatePassword, { loading, error }] = useMutation(UPDATE_PASSWORD);

  const messageHandler = (event, reason) => {
    if (reason === 'clickaway') return;
    setmessageState({ open: false });
  };

  const handleFormReset = () => {
    formApiRef.current.reset();
    formApiRef.current.setError('password', '');
  };

  const handleSubmit = async (data, callBack) => {
    try {
      setloadingState(true);

      const response = await callBack({
        variables: { ...data },
      });
      if (response.data) {
        setmessageState({
          message: intl.formatMessage(messages.passwordUpdated),
          open: true,
          type: 'success',
        });
        const timeOut = loadingAction(setmessageState, setloadingState);

        // Needed to prevent memory leak when component unmounts
        clearTimeout(timeOut);
        handleFormReset();
      }
    } catch (error) {
      const { graphQLErrors, networkError } = error;
      setloadingState(false);

      if (graphQLErrors.length > 0) {
        const [{ message }] = graphQLErrors;
        const invalidPassword = message.includes('Not authorized');

        setmessageState({
          message: invalidPassword
            ? intl.formatMessage(messages.invalidPassword)
            : message,
          open: true,
          type: 'error',
        });
      }

      if (networkError) {
        const { message } = networkError;
        const failedConnection = message.includes('Failed to fetch');
        if (failedConnection) {
          setmessageState({
            message: intl.formatMessage(globalMessages.networkError),
            open: true,
            type: 'error',
          });
        }
      }
    }
  };

  if (loading) return <Loading />;
  if (error) {
    return generalErrorHandler(error, { ...rest });
  }

  return (
    <>
      <ViewHeader
        component="Password Settings"
        body={intl.formatMessage(messages.pwdSettingsHeader)}
      />

      <div className={classes.root}>
        <Grid container>
          <Notification
            status={messageState.open}
            message={messageState.message}
            classes={classes}
            messageHandler={messageHandler}
            duration={3000}
            position={{ vertical: 'top', horizontal: 'right' }}
            type={messageState.type}
          />
          <Grid item xs={9} xl={10} className={classes.settingsContainer}>
            <InputWrapper>
              <Form
                getApi={formApi => {
                  formApiRef.current = formApi;
                }}
                id="change-password"
                onSubmit={formData => handleSubmit(formData, updatePassword)}
              >
                <Content
                  subHeader={intl.formatMessage(messages.newPassword)}
                  body={<NewPassword />}
                  containerClass={pwdSettingsClasses.newPassword}
                  headerClass={classes.sectionHeader}
                />
                <Content
                  subHeader={intl.formatMessage(messages.confirmChanges)}
                  body={<ConfirmChanges />}
                  containerClass={pwdSettingsClasses.confirmContainer}
                  headerClass={classes.sectionHeader}
                />
              </Form>
            </InputWrapper>
          </Grid>
          <ActionPanel
            action="update"
            type="submit"
            formId="change-password"
            resetForm={handleFormReset}
            loading={loadingState}
            disabled={loadingState}
          />
        </Grid>
      </div>
    </>
  );
}

PasswordSettings.propTypes = {
  intl: intlShape.isRequired,
};

const intlPasswordSettings = injectIntl(PasswordSettings);

export default intlPasswordSettings;
