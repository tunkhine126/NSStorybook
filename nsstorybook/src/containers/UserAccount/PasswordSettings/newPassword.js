/**
 *
 * NewPassword
 *
 */

import React from 'react';
import { FormattedMessage } from 'react-intl';

import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';

import InputValidation from 'components/shared/InputValidation';

import { required } from 'utils/validations';
import messages from '../messages';
import { commonStyles } from '../styles';

export function NewPassword() {
  const classes = commonStyles();

  const handleConfirmChange = (value, field, fieldApi) => {
    fieldApi.validate('passwordConfirmation');
  };
  const minCharacters = value => {
    if (value && value.length < 6) {
      return <FormattedMessage {...messages.minCharacters} />;
    }
    return undefined;
  };

  const passwordMatch = (value, values) =>
    values.password !== values.passwordConfirmation ? (
      <FormattedMessage {...messages.passwordMatch} />
    ) : (
      undefined
    );

  const combinedValidation = (value, values) =>
    required(value) || minCharacters(value) || passwordMatch(value, values);

  return (
    <>
      <Grid item className={classes.spacing}>
        <Typography variant="body1" className={classes.darkText} gutterBottom>
          <FormattedMessage {...messages.passwordReq} />
        </Typography>

        <InputValidation
          field="password"
          type="password"
          label={<FormattedMessage {...messages.newPassword} />}
          validate={combinedValidation}
          validateOnChange
          validateOnBlur
          notify={['passwordConfirmation']}
        />
        <InputValidation
          field="passwordConfirmation"
          type="password"
          label={<FormattedMessage {...messages.confirmPassword} />}
          validate={combinedValidation}
          onChange={handleConfirmChange}
          validateOnChange
          validateOnBlur
          notify={['password']}
        />
      </Grid>
    </>
  );
}

export default NewPassword;
