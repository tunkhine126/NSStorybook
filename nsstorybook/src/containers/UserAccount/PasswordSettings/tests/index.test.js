/**
 *
 * Tests for PasswordSettings
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from '@testing-library/react';

import TestingComponent from 'components/shared/Testing';
import PasswordSettings from '../Loadable';

describe('<PasswordSettings />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <TestingComponent>
        <PasswordSettings />
      </TestingComponent>
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <TestingComponent>
        <PasswordSettings />
      </TestingComponent>
    );
    expect(firstChild).toMatchSnapshot();
  });
});
