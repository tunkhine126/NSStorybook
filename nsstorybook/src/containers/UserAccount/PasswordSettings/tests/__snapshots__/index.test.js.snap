// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordSettings /> Should render and match the snapshot 1`] = `
<h2
  class="MuiTypography-root makeStyles-header-299 MuiTypography-h1 MuiTypography-gutterBottom"
>
  Change Your Password
</h2>
`;
